LFS-1147: The test runmode should have HANCESTRO already installed

IntegralProgrammer opened the pull request:
This PR adds the automatic installation of the HANCESTRO vocabulary into CARDS from BioPortal when `./start_cards.sh` is called with the `test` Sling runMode.

To test:
- Build this branch (`mvn clean install`)
- Start CARDS without specifying the `test` runMode (`./start_cards.sh` or `./start_cards.sh -Dsling.run.modes=dev,lfs`). CARDS should start and operate normally.
- Stop CARDS and cleanup (CTRL+C, `rm -rf sling`)
- Start CARDS with the `test` runMode but with no BioPortal API Key specified. (`./start_cards.sh -Dsling.run.modes=dev,test`). CARDS should start but the HANCESTRO vocabulary will not be installed. A `BIOPORTAL_APIKEY not specified` message should be printed to the console
- Stop CARDS and cleanup (CTRL+C, `rm -rf sling`)
- Start CARDS with the `test` runMode and an invalid BioPortal API Key (`export BIOPORTAL_APIKEY=1234` then `./start_cards.sh -Dsling.run.modes=dev,test`). CARDS should start but the HANCESTRO vocabulary will not be installed. A `HANCESTRO install failed` message should be printed to the console
- Stop CARDS and cleanup (CTRL+C, `rm -rf sling`)
- Start CARDS with the `test` runMode and with a _valid_ BioPortal API Key (`export BIOPORTAL_APIKEY=some-valid-api-key` then `./start_cards.sh -Dsling.run.modes=dev,test`). CARDS should start and operate normally in _test_ mode with the _HANCESTRO_ vocabulary installed.

IntegralProgrammer committed:
LFS-1147: The test runmode should have HANCESTRO already installed

sdumitriu commented:
One minor issue is that every time the instance is restarted, HANCETRO will be installed again. Can you first query `$CARDS_URL/Vocabularies/HANCESTRO.json` and see if it exists?

sdumitriu commented:
@sashaandjic Works for me, make sure that you start it with the `start_cards.sh` script, not by running the jar directly.

```
./start_cards.sh -Dsling.run.modes=dev,test,lfs
```

sdumitriu commented:
And also that you are on the right branch.

sashaandjic commented:
> @sashaandjic Works for me, make sure that you start it with the `start_cards.sh` script, not by running the jar directly.
@sdumitriu  I did not use the start_cards.sh script. Deleted my silly comment.

sdumitriu commented:
And another thing: it takes a little bit until the vocabulary gets installed, wait until you see the green `* Installed HANCESTRO *` message in the console.

sashaandjic commented:
> And another thing: it takes a little bit until the vocabulary gets installed, wait until you see the green `* Installed HANCESTRO *` message in the console.
I see now. Thanks!
![image](https://user-images.githubusercontent.com/23243244/117745689-f7993f80-b1d8-11eb-8342-c5d42ba5d544.png)

sashaandjic commented:
cards 4 kids
![image](https://user-images.githubusercontent.com/23243244/117749106-303c1780-b1df-11eb-8933-b76da2b65460.png)

sdumitriu reviewed (changes requested):
- make curl quiet
- check if HANCESTRO is not already installed before installing it
- print "cards ready" only after "hancestro installed", since it is not really ready before that

IntegralProgrammer committed:
LFS-1147: The test runmode should have HANCESTRO already installed

Make cURL quiet

IntegralProgrammer committed:
LFS-1147: The test runmode should have HANCESTRO already installed

Check if HANCESTRO is not already installed before installing it

IntegralProgrammer committed:
LFS-1147: The test runmode should have HANCESTRO already installed

Print 'Started CARDS' only after 'HANCESTRO installed' since it is not really ready before that

IntegralProgrammer committed:
LFS-1147: The test runmode should have HANCESTRO already installed

Always check if CARDS has started (using info.sessionInfo.json check) regardless of the runMode

IntegralProgrammer commented:
>     * make curl quiet
> 
>     * check if HANCESTRO is not already installed before installing it
> 
>     * print "cards ready" only after "hancestro installed", since it is not really ready before that

Addressed these comments and moved the `info.sessionInfo.json` check so that it is always executed, in all runModes, and not just the `test` runMode.

sashaandjic reviewed (approved):
tested - all suggested changes have been addressed

sdumitriu reviewed (approved):
I wonder if it's better to print `Waiting for CARDS to start` every 5 seconds or just once. Repeating it at least gives an indication to the user that something is still going on, but I'm not sure repetition is OK.

Due to the way endpoints become ready, it is possible to get started without HANCESTRO installed, reinstall, or even a double false negative (HANCESTRO appears not installed, but installing it fails). If the timing is "right", the session info becomes available before the content nodes become available, leading to `/Vocabularies/HANCESTRO` returning a 403, or the vocabulary indexing servlet may not be ready yet, in which case the request to install is not available. Not a problem with this PR, it would need toying with the startup ready detection mechanism.
